### PR TITLE
Cherry pick community stable/mitaka fixes for L3/OVS agents

### PR DESCRIPTION
**Pull request by jckasper**
Updating agents table is contantious operation which
can fail often if mysql backend is in multimaster mode.
This could lead to agents flapping and various issues
such as sporadic reschedluing, port binding failures, etc.

Change-Id: Ief392f9a09d86c185dc086055d2cbc1891ff1d7f
Closes-Bug: #1560724
(cherry picked from commit d5e4013556c7144d347ece267c4bd3c8dc87b24f)

Conflicts:
    neutron/db/agents_db.py

### CODE REVIEW

**Comment by jckasper** [comment]
This is stable/mitaka cherry pick: 1 of 7.

**Comment by jckasper** [comment]
This is stable/mitaka cherry pick: 2 of 7.

**Comment by rtheis** [comment]
These cherry picks look good to me.  Could these be cherry picked upstream too?

**Comment by jckasper** [comment]
These two... maybe.   They are not security fixes so I'm not sure if the community would want to cherry pick them back to liberty or not.

**Comment by jckasper** [comment]
Stable/mitaka cherry pick: 3 of 7  has been added.   As you can see this is a much larger change.   Each of these cherry picks should be a separate pull request, but I'm not sure how to do that.

File: neutron/plugins/ml2/drivers/openvswitch/agent/ovs_neutron_agent.py [hunk 1]
```diff
@@ -183,7 +185,7 @@ def __init__(self, bridge_classes, integ_br, tun_br, local_ip,
         super(OVSNeutronAgent, self).__init__()
         self.conf = conf or cfg.CONF
 
-        self.fullsync = True
+        self.fullsync = False
```
Comment (rtheis) [inline neutron/plugins/ml2/drivers/openvswitch/agent/ovs_neutron_agent.py:182]: Why was this changed?

**Comment by rtheis** [comment]
I added one question to cherry pick 3.

File: neutron/plugins/ml2/drivers/openvswitch/agent/ovs_neutron_agent.py (hunk 1 above)
Comment (jckasper) [inline neutron/plugins/ml2/drivers/openvswitch/agent/ovs_neutron_agent.py:182]: That is what was in patch 3 of 7  (along with lots of other changes too).   New logic probably does not want that initialized to True because it needs to finish setting up objects before sync is done.

Comment (jckasper) [inline neutron/plugins/ml2/drivers/openvswitch/agent/ovs_neutron_agent.py:182]: self.fullsync = True    is done on L336 after reporting state of the agent to neutron server.   So.....  don't want the fullsync to be done until after agent has reported into the server.

**Comment by jckasper** [comment]
This pull request is going to get much larger...  as I pull in the remaining 4 patches.  3 of the 4 patches build on the changes done in this latest patch so it is probably good to keep all of these together.

File: neutron/plugins/ml2/drivers/openvswitch/agent/ovs_neutron_agent.py (hunk 1 above)
Comment (rtheis) [inline neutron/plugins/ml2/drivers/openvswitch/agent/ovs_neutron_agent.py:182]: Thanks for the explanation.

**Comment by rtheis** [comment]
Cherry pick 3 lgtm

**Comment by jckasper** [comment]
Stable/mitaka cherry pick: 4 of 7 has been added.  Small change this time.

**Comment by jckasper** [comment]
Stable/mitaka cherry pick: 5 of 7 has been added.

**Comment by rtheis** [comment]
Cherry pick 4 lgtm

**Comment by jckasper** [comment]
Stable/mitaka cherry pick: 6 of 7 has been added.   One left ....

**Comment by jckasper** [comment]
Final....  stable/mitaka cherry pick has been added.   7 of 7.    Richard, thank you for the reviews.

**Comment by rtheis** [comment]
Cherry pick 6 lgtm

**Comment by rtheis** [comment]
Cherry pick 7 lgtm

**Comment by rtheis** [comment]
Cherry pick 5 lgtm
